homeassistant/components/tesla/__init__.py

```diff
@@ -261,16 +301,20 @@ def device_info(self):
 
     async def async_added_to_hass(self):
         """Register state update callback."""
+        self.async_on_remove(self.coordinator.async_add_listener(self.refresh))
 
     async def async_will_remove_from_hass(self):
         """Prepare for unload."""
 
     async def async_update(self):
         """Update the state of the device."""
-        if self.controller.is_token_refreshed():
-            (refresh_token, access_token) = self.controller.get_tokens()
-            _async_save_tokens(
-                self.hass, self.config_entry, access_token, refresh_token
-            )
-            _LOGGER.debug("Saving new tokens in config_entry")
-        await self.tesla_device.async_update()
+        await self.coordinator.async_request_refresh()
+        self.refresh()
```

Comment:
i feel tbis might be doing double requests here because you listen for the coordinator update and call refresh when that update takes place.

Reply:
Actually it doesn't. I had to redo the underlying API so we could have a synchronous callback.  The update in the coordinator tells the teslajsonpy API to update its central cache from Tesla's servers. The refresh is having the teslajsonpy API then update its entities with that central cache. Previously the update function did both.

My understanding is `async_update` isn't actually called anymore since polling is false and really is here just for legacy purposes.

Reply:
async_update is actually possible to call via update entity service so ghats why you see components adding support for triggering update request

Reply:
Ok makes sense. That service call should still work if called but of course will be debounced by the DataUpdateCoordinator.

Reply:
This should be removed, since the coordinator will signal the entity to update after the data is refreshed.

Reply:
You mean async_update should be removed?

Reply:
No only line 312.

Reply:
Currently line 315.